Define interfaces for consensus

mrBliss opened the pull request:
In the `Shelley.Spec.Ledger.API.*` modules, define type classes for the most
important functionality consensus. These classes are parametric in the era.
Instances should be provided for each era (Shelley, Allegra, Mary, etc.). At the
moment, only instances for `ShelleyEra` are provided.

These classes list a bunch of super-class constraints that consensus relies on.
I have tried to be complete, but there's a good chance I forgot to add a few
constraints that are currently already satisfied.

Default implementations are provided for each method so that an empty instance
declaration should be enough for each class + era combination. The constraints
on these instances are what matters, they should be minimal, e.g., just `Crypto
crypto` and `DSignable ..`.

Note: the super-class constraints of the classes are what consensus needs, *not*
what the default implementations of the methods need. If a default
implementation needs more constraints, add them to the *default signature*.

mrBliss committed:
Define interfaces for consensus

In the `Shelley.Spec.Ledger.API.*` modules, define type classes for the most
important functionality consensus. These classes are parametric in the era.
Instances should be provided for each era (Shelley, Allegra, Mary, etc.). At the
moment, only instances for `ShelleyEra` are provided.

These classes list a bunch of super-class constraints that consensus relies on.
I have tried to be complete, but there's a good chance I forgot to add a few
constraints that are currently already satisfied.

Default implementations are provided for each method so that an empty instance
declaration should be enough for each class + era combination. The constraints
on these instances are what matters, they should be minimal, e.g., just `Crypto
crypto` and `DSignable ..`.

Note: the super-class constraints of the classes are what consensus needs, *not*
what the default implementations of the methods need. If a default
implementation needs more constraints, add them to the *default signature*.

mrBliss commented:
Draft because this depends on #1902.

mrBliss commented:
#1902 has been merged. I have rebased this PR on master. Ready for review.

mrBliss commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Validation.hs:
```diff
@@ -31,18 +34,147 @@ import GHC.Generics (Generic)
+    NoThunks (BlockTransitionError era),
+    Eq (STS.PredicateFailure (STS.CHAIN era)),
+    Show (STS.PredicateFailure (STS.CHAIN era)),
+    NoThunks (STS.PredicateFailure (STS.CHAIN era))
```
This list is already rather long, but I expect some things are missing. So at some point, I might have to add more constraints here.

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Mempool.hs:
```diff
@@ -33,13 +31,57 @@ import Control.State.Transition.Extended
+import Shelley.Spec.Ledger.TxBody (EraIndependentTxBody)
+
+-- TODO #1304: add reapplyTxs
+class
```
I might move these class definitions to `Cardano.Ledger.API` since we now expect them to apply across eras. But can do that in a separate PR

mrBliss replied:
We decided to postpone this.

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Mempool.hs:
```diff
@@ -33,13 +31,57 @@ import Control.State.Transition.Extended
+    Show (Tx era),
+    NoThunks (Tx era),
+    FromCBOR (Annotator (Tx era)),
+    ToCBOR (Tx era),
```
Tim has proposed packaging these things up as `ChainData`, which would make this list somewhat cleaner.

mrBliss replied:
See the last commit.

nc6 commented on shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/API/Validation.hs:
```diff
@@ -31,18 +34,147 @@ import GHC.Generics (Generic)
+  -- few header level checks, such as size constraints.
+  applyTick ::
+    Globals ->
+    ShelleyState era ->
```
[bikeshedding] `ShelleyState` was an alias for downstream use, but now feels a bit awkward, since generally this won't be shelley. We could use the wrapped type `NewEpochState`, or come up with a better name (but all the obvious ones are taken...)

mrBliss replied:
I have gone for `NewEpochState` for now.

mrBliss committed:
Remove the ShelleyState type synonym for NewEpochState

It will be used by all eras, so having Shelley in the name is confusing.

mrBliss committed:
Introduce ChainData, SerialisableData, and AnnotatedData

These constraint synonyms make the contexts of the API classes much more
readable.
